compiler/dyno/lib/util/filesystem.cpp

```diff
@@ -115,4 +150,67 @@ bool fileExists(const char* path) {
 }
 
 
+std::error_code deleteDir(std::string dirname) {
+  // LLVM 5 added remove_directories
+  return llvm::sys::fs::remove_directories(dirname, false);
 }
+
+std::error_code makeTempDir(std::string dirPrefix, std::string& tmpDirPathOut) {
+  std::string tmpdirprefix = std::string(getTempDir()) + "/" + dirPrefix;
+  std::string tmpdirsuffix = ".deleteme-XXXXXX";
+
+  struct passwd* passwdinfo = getpwuid(geteuid());
+  const char* userid;
+  if (passwdinfo == NULL) {
+    userid = "anon";
+  } else {
+    userid = passwdinfo->pw_name;
+  }
+  char* myuserid = strdup(userid);
+  removeSpacesBackslashesFromString(myuserid);
+  std::string tmpDir = tmpdirprefix + std::string(myuserid) + tmpdirsuffix;
+  char* tmpDirMut = strdup(tmpDir.c_str());
+  // TODO: we could use llvm::sys::fs::createUniqueDirectory instead of mkdtemp
+  // see the comment in LLVM source
+  // https://llvm.org/doxygen/Path_8cpp_source.html#l00883
+  auto dirRes = std::string(mkdtemp(tmpDirMut));
+
+  // get the posix error code if mkdtemp failed.
+  int err = errno;
+
+  if (dirRes.empty()) {
+    std::error_code ret = make_error_code(static_cast<std::errc>(err));
+    return ret;
+  }
+
+  free(myuserid); myuserid = NULL;
+  free(tmpDirMut); tmpDirMut = NULL;
+
+  tmpDirPathOut = dirRes;
+  return std::error_code();
+}
+
+std::error_code ensureTmpDirExists(std::string& saveCDir,
```

Comment:
I don't think that this should live in dyno, personally. The formals seem too tied up in the specifics of the old compiler driver. Could we make it a static function there?

Reply:
I moved it back out